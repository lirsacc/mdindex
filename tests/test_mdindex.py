import contextlib
import itertools
import shutil
import subprocess
import tempfile
from collections.abc import Iterator
from pathlib import Path

EXAMPLES_DIR = Path(__file__).parent / "examples"
TOC_EXAMPLES_DIR = Path(__file__).parent / "examples-toc"


@contextlib.contextmanager
def workdir(src: Path) -> Iterator[Path]:
    with tempfile.TemporaryDirectory() as tempdir:
        shutil.copytree(src, tempdir, dirs_exist_ok=True)
        yield Path(tempdir)


# Smoke test the base copy worked as expected
def test_smoke() -> None:
    with workdir(EXAMPLES_DIR) as test_dir:
        assert take_snapshot(EXAMPLES_DIR) == take_snapshot(test_dir)


def take_snapshot(ref: Path) -> str:
    return "\n".join(
        itertools.chain.from_iterable(
            [
                f">>>> START FILE: {x.relative_to(ref)}",
                x.read_text(),
                "<<<< END FILE",
            ]
            for x in ref.glob("**/*.md")
        )
    )


def test_examples(snapshot: object) -> None:
    with workdir(EXAMPLES_DIR) as test_dir:
        subprocess.run(
            ["python", "-m", "mdindex", str(test_dir), "-r"],
            cwd=test_dir,
            check=True,
            capture_output=True,
        )

        assert snapshot == take_snapshot(test_dir)


def test_examples_no_recursive(snapshot: object) -> None:
    with workdir(EXAMPLES_DIR) as test_dir:
        subprocess.run(
            ["python", "-m", "mdindex", str(test_dir)],
            cwd=test_dir,
            check=True,
            capture_output=True,
        )

        assert snapshot == take_snapshot(test_dir)


def test_ignore(snapshot: object) -> None:
    with workdir(EXAMPLES_DIR) as test_dir:
        subprocess.run(
            [
                "python",
                "-m",
                "mdindex",
                str(test_dir / "foo"),
                "-r",
                # Will ignore everything under deep
                "--ignore=**/deep/**",
                # Will not ignore two.md as that's not a match
                "--ignore=two",
            ],
            cwd=test_dir / "foo",
            check=True,
            capture_output=True,
        )

        assert snapshot == take_snapshot(test_dir / "foo")


def test_toc(snapshot: object) -> None:
    with workdir(TOC_EXAMPLES_DIR) as test_dir:
        subprocess.run(
            [
                "python",
                "-m",
                "mdindex",
                str(test_dir),
                "-r",
                "--toc",
            ],
            cwd=test_dir,
            check=True,
            capture_output=True,
        )

        assert snapshot == take_snapshot(test_dir)
