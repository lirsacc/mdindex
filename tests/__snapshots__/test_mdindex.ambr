# serializer version: 1
# name: test_examples
  '''
  >>>> START FILE: root.md
  # Root file
  
  <<<< END FILE
  >>>> START FILE: README.md
  # Examples for testing
  
  <!-- mdindex:index:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [Root file](./root.md)
  - [Examples: bar/README](./bar)
    - [One](./bar/one.md)
    - [Two](./bar/two.md)
  - [Baz](./baz)
    - [Camel Cased Filename](./baz/camelCasedFilename.md)
    - [Dash Cased Filename](./baz/dash-cased-filename.md)
    - [Snake Cased Filename Copy](./baz/snake_cased_filename%20copy.md)
  - [Foo](./foo)
    - [This file should be included with this title](./foo/one.md)
    - [This file should be included with this title too](./foo/two.md)
    - [This uses an underline title which should also work](./foo/underlined.md)
    - [Deep](./foo/deep)
      - [Nested1](./foo/deep/nested1.md)
      - [Deeper](./foo/deep/deeper)
  
  <!-- mdindex:index:end -->
  
  ---
  
  This file should end up containing a summary above this.
  
  <<<< END FILE
  >>>> START FILE: foo/README.md
  <!-- mdindex:index:start -->
  # Foo
  
  <!-- WARN: This file is auto-generated. Do not edit manually. -->
  
  - [This file should be included with this title](./one.md)
  - [This file should be included with this title too](./two.md)
  - [This uses an underline title which should also work](./underlined.md)
  - [Deep](./deep)
    - [Nested1](./deep/nested1.md)
    - [Deeper](./deep/deeper)
      - [Nested2](./deep/deeper/nested2.md)
      - [Too Deep](./deep/deeper/too-deep)
  <!-- mdindex:index:end -->
  
  <<<< END FILE
  >>>> START FILE: foo/one.md
  # This file should be included with this title
  
  Content after the title doesn't matter.
  
  <<<< END FILE
  >>>> START FILE: foo/two.md
  Content before the title doesn't matter we pick the first one.
  
  # This file should be included with this title too
  
  Content after the title still doesn't matter.
  
  <<<< END FILE
  >>>> START FILE: foo/underlined.md
  This uses an underline title which should also work
  ===================================================
  
  <<<< END FILE
  >>>> START FILE: baz/dash-cased-filename.md
  This file doesn't have a title in the file so we'll include the title inferred from the filename, parsing dash case into title case.
  
  <<<< END FILE
  >>>> START FILE: baz/snake_cased_filename copy.md
  This file doesn't have a title in the file so we'll include the title inferred from the filename, parsing snake case into title case.
  
  <<<< END FILE
  >>>> START FILE: baz/README.md
  <!-- mdindex:index:start -->
  # Baz
  
  <!-- WARN: This file is auto-generated. Do not edit manually. -->
  
  - [Camel Cased Filename](./camelCasedFilename.md)
  - [Dash Cased Filename](./dash-cased-filename.md)
  - [Snake Cased Filename Copy](./snake_cased_filename%20copy.md)
  <!-- mdindex:index:end -->
  
  <<<< END FILE
  >>>> START FILE: baz/camelCasedFilename.md
  This file doesn't have a title in the file so we'll include the title inferred from the filename, parsing camel case into title case.
  
  <<<< END FILE
  >>>> START FILE: bar/README.md
  # Examples: bar/README
  
  This file should not end up containing a summary as it exists already without markers.
  
  <<<< END FILE
  >>>> START FILE: bar/one.md
  
  <<<< END FILE
  >>>> START FILE: bar/two.md
  
  <<<< END FILE
  >>>> START FILE: foo/deep/nested1.md
  
  <<<< END FILE
  >>>> START FILE: foo/deep/README.md
  <!-- mdindex:index:start -->
  # Deep
  
  <!-- WARN: This file is auto-generated. Do not edit manually. -->
  
  - [Nested1](./nested1.md)
  - [Deeper](./deeper)
    - [Nested2](./deeper/nested2.md)
    - [Too Deep](./deeper/too-deep)
      - [Nested3](./deeper/too-deep/nested3.md)
  <!-- mdindex:index:end -->
  
  <<<< END FILE
  >>>> START FILE: foo/deep/deeper/README.md
  <!-- mdindex:index:start -->
  # Deeper
  
  <!-- WARN: This file is auto-generated. Do not edit manually. -->
  
  - [Nested2](./nested2.md)
  - [Too Deep](./too-deep)
    - [Nested3](./too-deep/nested3.md)
  <!-- mdindex:index:end -->
  
  <<<< END FILE
  >>>> START FILE: foo/deep/deeper/nested2.md
  
  <<<< END FILE
  >>>> START FILE: foo/deep/deeper/too-deep/README.md
  <!-- mdindex:index:start -->
  # Too Deep
  
  <!-- WARN: This file is auto-generated. Do not edit manually. -->
  
  - [Nested3](./nested3.md)
  <!-- mdindex:index:end -->
  
  <<<< END FILE
  >>>> START FILE: foo/deep/deeper/too-deep/nested3.md
  
  <<<< END FILE
  '''
# ---
# name: test_examples_no_recursive
  '''
  >>>> START FILE: root.md
  # Root file
  
  <<<< END FILE
  >>>> START FILE: README.md
  # Examples for testing
  
  <!-- mdindex:index:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [Root file](./root.md)
  - [Examples: bar/README](./bar)
    - [One](./bar/one.md)
    - [Two](./bar/two.md)
  - [Baz](./baz)
    - [Camel Cased Filename](./baz/camelCasedFilename.md)
    - [Dash Cased Filename](./baz/dash-cased-filename.md)
    - [Snake Cased Filename Copy](./baz/snake_cased_filename%20copy.md)
  - [Foo](./foo)
    - [This file should be included with this title](./foo/one.md)
    - [This file should be included with this title too](./foo/two.md)
    - [This uses an underline title which should also work](./foo/underlined.md)
    - [Deep](./foo/deep)
      - [Nested1](./foo/deep/nested1.md)
      - [Deeper](./foo/deep/deeper)
  
  <!-- mdindex:index:end -->
  
  ---
  
  This file should end up containing a summary above this.
  
  <<<< END FILE
  >>>> START FILE: foo/one.md
  # This file should be included with this title
  
  Content after the title doesn't matter.
  
  <<<< END FILE
  >>>> START FILE: foo/two.md
  Content before the title doesn't matter we pick the first one.
  
  # This file should be included with this title too
  
  Content after the title still doesn't matter.
  
  <<<< END FILE
  >>>> START FILE: foo/underlined.md
  This uses an underline title which should also work
  ===================================================
  
  <<<< END FILE
  >>>> START FILE: baz/dash-cased-filename.md
  This file doesn't have a title in the file so we'll include the title inferred from the filename, parsing dash case into title case.
  
  <<<< END FILE
  >>>> START FILE: baz/snake_cased_filename copy.md
  This file doesn't have a title in the file so we'll include the title inferred from the filename, parsing snake case into title case.
  
  <<<< END FILE
  >>>> START FILE: baz/camelCasedFilename.md
  This file doesn't have a title in the file so we'll include the title inferred from the filename, parsing camel case into title case.
  
  <<<< END FILE
  >>>> START FILE: bar/README.md
  # Examples: bar/README
  
  This file should not end up containing a summary as it exists already without markers.
  
  <<<< END FILE
  >>>> START FILE: bar/one.md
  
  <<<< END FILE
  >>>> START FILE: bar/two.md
  
  <<<< END FILE
  >>>> START FILE: foo/deep/nested1.md
  
  <<<< END FILE
  >>>> START FILE: foo/deep/deeper/nested2.md
  
  <<<< END FILE
  >>>> START FILE: foo/deep/deeper/too-deep/nested3.md
  
  <<<< END FILE
  '''
# ---
# name: test_ignore
  '''
  >>>> START FILE: README.md
  <!-- mdindex:index:start -->
  # Foo
  
  <!-- WARN: This file is auto-generated. Do not edit manually. -->
  
  - [This file should be included with this title](./one.md)
  - [This file should be included with this title too](./two.md)
  - [This uses an underline title which should also work](./underlined.md)
  <!-- mdindex:index:end -->
  
  <<<< END FILE
  >>>> START FILE: one.md
  # This file should be included with this title
  
  Content after the title doesn't matter.
  
  <<<< END FILE
  >>>> START FILE: two.md
  Content before the title doesn't matter we pick the first one.
  
  # This file should be included with this title too
  
  Content after the title still doesn't matter.
  
  <<<< END FILE
  >>>> START FILE: underlined.md
  This uses an underline title which should also work
  ===================================================
  
  <<<< END FILE
  >>>> START FILE: deep/nested1.md
  
  <<<< END FILE
  >>>> START FILE: deep/deeper/nested2.md
  
  <<<< END FILE
  >>>> START FILE: deep/deeper/too-deep/nested3.md
  
  <<<< END FILE
  '''
# ---
# name: test_toc
  '''
  >>>> START FILE: multiple-h1.md
  Multiple h1 are technically wrong but we should handle gracefully.
  
  <!-- mdindex:toc:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [multiple-h1](#multiple-h1)
    - [Some h2 title](#some-h2-title)
      - [Some h3 title](#some-h3-title)
        - [Some h4 title](#some-h4-title)
  - [Some other h1](#some-other-h1)
    - [Some underlined h2 title](#some-underlined-h2-title)
      - [Some h4 title at the wrong nesting level](#some-h4-title-at-the-wrong-nesting-level)
  
  <!-- mdindex:toc:end -->
  
  # multiple-h1
  
  ## Some h2 title
  
  ### Some h3 title
  
  #### Some h4 title
  
  Some other h1
  =====
  
  Some underlined h2 title
  ---------
  
  #### Some h4 title at the wrong nesting level
  
  <<<< END FILE
  >>>> START FILE: h1-underlined.md
  h1-underlined
  ==========
  
  <!-- mdindex:toc:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [Some h2 title](#some-h2-title)
    - [Some h3 title](#some-h3-title)
      - [Some h4 title](#some-h4-title)
  - [Some underlined h2 title](#some-underlined-h2-title)
    - [Some h4 title at the wrong nesting level](#some-h4-title-at-the-wrong-nesting-level)
  
  <!-- mdindex:toc:end -->
  
  This toc should still be well formatted without an h1 title and inserted just
  after the first h1 title.
  
  ## Some h2 title
  
  ### Some h3 title
  
  #### Some h4 title
  
  Some underlined h2 title
  ---------
  
  #### Some h4 title at the wrong nesting level
  
  <<<< END FILE
  >>>> START FILE: no-h1.md
  This toc should still be well formatted without an h1 title and inserted just
  before the first h2 title.
  
  <!-- mdindex:toc:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [Some h2 title](#some-h2-title)
    - [Some h3 title](#some-h3-title)
      - [Some h4 title](#some-h4-title)
  - [Some underlined h2 title](#some-underlined-h2-title)
    - [Some h4 title at the wrong nesting level](#some-h4-title-at-the-wrong-nesting-level)
  
  <!-- mdindex:toc:end -->
  
  ## Some h2 title
  
  ### Some h3 title
  
  #### Some h4 title
  
  Some underlined h2 title
  ---------
  
  #### Some h4 title at the wrong nesting level
  
  <<<< END FILE
  >>>> START FILE: h1.md
  # h1
  
  <!-- mdindex:toc:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [Some h2 title](#some-h2-title)
    - [Some h3 title](#some-h3-title)
      - [Some h4 title](#some-h4-title)
  - [Some underlined h2 title](#some-underlined-h2-title)
    - [Some h4 title at the wrong nesting level](#some-h4-title-at-the-wrong-nesting-level)
  
  <!-- mdindex:toc:end -->
  
  This toc should still be well formatted without an h1 title and inserted just
  after the first h1 title.
  
  ## Some h2 title
  
  ### Some h3 title
  
  #### Some h4 title
  
  Some underlined h2 title
  ---------
  
  #### Some h4 title at the wrong nesting level
  
  <<<< END FILE
  >>>> START FILE: README.md
  # TOC examples
  
  <!-- mdindex:index:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [h1-underlined](./h1-underlined.md)
  - [h1](./h1.md)
  - [multiple-h1](./multiple-h1.md)
  - [Some h3 title](./no-h1-or-h2.md)
  - [Some h2 title](./no-h1.md)
  - [No Headings](./no-headings.md)
  - [Standard TOC](./standard.md)
  - [Too few headings](./too-few-headings.md)
  
  <!-- mdindex:index:end -->
  
  <<<< END FILE
  >>>> START FILE: too-few-headings.md
  # Too few headings
  
  This file has too few headings and thus no toc will be generated
  
  <!-- mdindex:toc:start -->
  <!-- mdindex:toc:end -->
  
  ## Some h2 title
  
  <<<< END FILE
  >>>> START FILE: no-headings.md
  This file has no headings and will have not TOC
  
  <!-- mdindex:toc:start -->
  <!-- mdindex:toc:end -->
  
  <<<< END FILE
  >>>> START FILE: standard.md
  # Standard TOC
  
  This toc should be well formatted despite nesting issues.
  
  <!-- mdindex:toc:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [Some h2 title](#some-h2-title)
    - [Some h3 title](#some-h3-title)
      - [Some h4 title](#some-h4-title)
  - [Some underlined h2 title](#some-underlined-h2-title)
    - [Some h4 title at the wrong nesting level](#some-h4-title-at-the-wrong-nesting-level)
  
  <!-- mdindex:toc:end -->
  
  ## Some h2 title
  
  ### Some h3 title
  
  #### Some h4 title
  
  Some underlined h2 title
  ---------
  
  #### Some h4 title at the wrong nesting level
  
  <<<< END FILE
  >>>> START FILE: no-h1-or-h2.md
  This toc should still be well formatted without an h1 title and inserted just
  before the first h2 title.
  
  <!-- mdindex:toc:start -->
  <!-- WARN: This section is auto-generated. Do not edit manually. -->
  
  - [Some h3 title](#some-h3-title)
    - [Some h4 title](#some-h4-title)
    - [Some h4 title at the wrong nesting level](#some-h4-title-at-the-wrong-nesting-level)
  - [Some other h3 title](#some-other-h3-title)
  
  <!-- mdindex:toc:end -->
  
  ### Some h3 title
  
  #### Some h4 title
  
  #### Some h4 title at the wrong nesting level
  
  ### Some other h3 title
  
  <<<< END FILE
  '''
# ---
